src/Control/Monad/Trans/MSF/Maybe.hs

```diff
@@ -65,11 +66,14 @@ catchMaybe msf1 msf2 = safely $ do
 
 -- * Converting to and from 'MaybeT'
 
+exceptToMaybeS :: (Functor m, Monad m) => MSF (ExceptT e m) a b -> MSF (MaybeT m) a b
+exceptToMaybeS = morphS $ MaybeT . fmap (either (const Nothing) Just) . runExceptT
+
 -- | Converts a list to an 'MSF' in 'MaybeT',
 --   which outputs an element of the list at each step,
 --   throwing 'Nothing' when the list ends.
-listToMaybeS :: Monad m => [b] -> MSF (MaybeT m) a b
-listToMaybeS = foldr iPost exit
+listToMaybeS :: (Functor m, Monad m) => [a] -> MSF (MaybeT m) arbitrary a
+listToMaybeS = exceptToMaybeS . runMSFExcept . sequence . map (step . const . return . (, ()))
```

Comment:
What is the type of: `exceptToMaybeS . runMSFExcept`?
What is the type of `sequence . map (step . const . return . (, ()))`, and what does it do?

Reply:
So, when I try to think about this function, it helps me to group this:
```
runMSFExcept . sequence . map (step . const . return . ((\x -> ((), x))))
  :: Monad m => [a1] -> MSF (ExceptT [a1] m) a2 ()
```
What I do not fully understand is why is conceptually correct. This seems to throw an exception with the type of the list at the end.

Reply:
The exception values are discarded. I'll document that.

> What is the type of `sequence . map (step . const . return . (, ()))`, and what does it do?

It outputs each list element and then throws an exception. I could refactor it as a function in `Control.Monad.Trans.MSF.Except` if you want.

Reply:
- `sequence . map` is `mapM`.

- Is sequence . map (step . const . return . (, ())) different from `once . const . return`?

Reply:
> I could refactor it as a function in `Control.Monad.Trans.MSF.Except` if you want.

No need. I'm just trying to understand it and document it. It may help to give names or types to local definitions in a `where` clause.

Reply:
>     * `sequence . map` is `mapM`.

Good point. Fixed. We can even take `mapM_`, to emphasize that the returned list is irrelevant, and only the effect of outputting the elements is important.
 
>     * Is `sequence . map (step . const . return . (, ()))` different from `once . const . return`?

Yes. `once . const . return` does one tick only. But we want to perform one tick for every list element.

Reply:
I have trouble understanding this. I get it, but I don't get it.

I tried to run it on my machine, and the result I got was the same.

Reply:
Does it help if I break it up into understandable pieces? This is a big function.

Reply:
It would, yes. Can you?

It's perfect if the "pieces" are in a where in the same definition. There's no need for now to make them top-level.

Reply:
Turns out I could find some good meaning for the pieces. I made them top level now for haddock, if that's ok.